### PR TITLE
Update pyenv pyenvinstall Make targets

### PR DESCRIPTION
**Pull request by dalf**
## What does this PR do?

Change about utils/makefile.python: 
* `make pyenv` makes sure that there is a python executable in the virtualenv, if not the virtualenv is installed again.
* `make pyenvinstall` makes to install `pyyaml` before running setup.py (setup.py requires pyyml to read searx/settings.yml)

## Why is this change important?

* github action workflow caches the ./local directory : 
https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/.github/workflows/integration.yml#L25-L32
* sometimes this cache can be corrupted (it is not clear how the cache gets corrupted).
* this PR ensure that `make test` works as expected even if
    * the python interpreter is missing
    * the virtualenv exists but pyyaml is missing

## How to test this PR locally?

```
make pyenvinstall
rm ./local/py3/bin/python
make test
source ./local/py3/bin/activate
pip uninstall pyyaml
make pyenvinstall
```

### CODE REVIEW

**Comment by dalf** [comment]
From https://github.com/searx/searx/pull/2514#issuecomment-769731867 :

> Do you know why it crashes now and how we can fix it?

The cache contains a copy './local'. 
It is not possible to clear the cache: https://github.com/actions/cache/issues/2#issuecomment-673493515 (see the top of the page)

It seems the cache is corrupted.
* the directory contains `./local` without proper initialization.
* so `make pyenv` does nothing since the `./local` exists.
* so `make pyenvinstall` fails because `pyyaml` is not installed (since `./setup.py` requires pyyaml)

**Comment by dalf** [comment]
It should fix the CI.

@return42  can you give a look ?

File: utils/makefile.python [hunk 1]
```diff
@@ -99,7 +99,8 @@ quiet_cmd_pyinstall = INSTALL   $2
 quiet_cmd_pyenvinstall = PYENV     install $2
       cmd_pyenvinstall = \
 	if ! cat $(PY_ENV)/requirements.sha256 2>/dev/null | sha256sum --check --status 2>/dev/null; then \
-		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS) ;\
+		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install $$(cat ./requirements.txt | grep -i pyyaml); \
```
Comment (return42) [inline utils/makefile.python]: This line can also be written:

     $$(grep -i pyyaml ./requirements.txt)

File: utils/makefile.python [hunk 2]
```diff
@@ -99,7 +99,8 @@ quiet_cmd_pyinstall = INSTALL   $2
...
+		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS) &&\
 		sha256sum requirements*.txt > $(PY_ENV)/requirements.sha256 ;\
```
Comment (return42) [inline utils/makefile.python:120]: OK, you only want to create a `$(PY_ENV)/requirements.sha256 ` when pip succeeds .. make sense to me.

Does it make sense to have a:

    rm -f $(PY_ENV)/requirements.sha256 ;

before pip installs?

File: utils/makefile.python [hunk 3]
```diff
@@ -119,13 +120,13 @@ quiet_cmd_pyenvuninstall = PYENV     uninstall   $2
 # $2 path to folder where virtualenv take place
 quiet_cmd_virtualenv  = PYENV     usage: $ source ./$@/bin/activate
       cmd_virtualenv  = \
-	if [ ! -d "./$(PY_ENV)" ];then                                  \
-		$(PYTHON) -m venv $(VTENV_OPTS) $2;                     \
+	if [ -d "./$(PY_ENV)" -a -x "./$(PY_ENV_BIN)/python" ]; then \
```
Comment (return42) [inline utils/makefile.python:139]: Checking the existence of `./$(PY_ENV_BIN)/python"` makes also sense to me.  This might help in cases where a previous build of the virtualenv has been failed for any reason.  

I thought twice about a more robust criteria but I don't have anything better to suggest.

**Review by return42 — DISMISSED** [review]
See my comments.  All in all it LGTM .. there are two things that bother me a bit.

1. I made some test and killed the `make clean pyenv` to get a corrupt `/local` .. the pyyaml problem still exists under some circumstance.
2. `makefile.python` is _abstract_, only and handles _common python, pip and setuptools issues_. This patch adds some `pyyaml` aspects and is not an _abstract_ solution.

At the moment I don't have more time and I also don't have a better solution right now.  Is it OK for you to wait with merge.  Otherwise I can also think about a more _abstract_ solution afterwards.

File: utils/makefile.python (hunk 2 above)
Comment (dalf) [reply utils/makefile.python:120]: Done in fe3085d68e965732ca7291a14225b34a7e046e92

**Comment by dalf** [comment]
> makefile.python is _abstract_

Abstract version:
* replace `$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install $$(grep pyyaml ./requirements.txt);`
* by `$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -r ./requirements-presetup.txt;`
* `requirements-presetup.txt` contains only one line `pyyaml==5.4.1` (to content has to be synchronized with `requirements.txt`).

But:
* If we could avoid to another requirements*.txt, it would be nice
* I prefer to find a way to remove the pyyaml dependency (related to https://github.com/searx/searx/pull/2408 )
* Is it a problem if utils/makefile.python is not abstract ?

File: utils/makefile.python (hunk 3 above)
Comment (return42) [reply utils/makefile.python:139]: Another thought:  now, with removing ` $(PY_ENV)/requirements.sha256` before running `pip install` and the `&& sha256sum requirements*.txt > $(PY_ENV)/requirements.sha256` from line 104-105, checking existence of  ` $(PY_ENV)/requirements.sha256`  could be a good criteria here?

File: utils/makefile.python [hunk 4]
```diff
@@ -99,7 +99,9 @@ quiet_cmd_pyinstall = INSTALL   $2
 quiet_cmd_pyenvinstall = PYENV     install $2
       cmd_pyenvinstall = \
 	if ! cat $(PY_ENV)/requirements.sha256 2>/dev/null | sha256sum --check --status 2>/dev/null; then \
-		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS) ;\
+		rm -f $(PY_ENV)/requirements.sha256; \
+		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install $$(grep pyyaml ./requirements.txt); \
+		$(PY_ENV_BIN)/python -m pip $(PIP_VERBOSE) install -e $2$(PY_SETUP_EXTRAS) &&\
```
Comment (return42) [inline utils/makefile.python:119]: Thought: if I ignore the git workflow caching issue and have a complete new look at this two lines, I cant understand why line 103 is needed .. or did I lost the overview?

Comment (dalf) [reply utils/makefile.python:119]: https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/setup.py#L11

brand is a variable initialized in `searx/__init__.py` :

https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/searx/__init__.py#L27

https://github.com/searx/searx/blob/71d66979c2935312e0aed7fc7c3cf6199fbe88a2/searx/settings_loader.py#L8

File: utils/makefile.python (hunk 3 above)
Comment (dalf) [reply utils/makefile.python:139]: Is there a case where the `pyenv` target is invoked but not the `pyenvinstall` target ?
* if no, why there are two targets ?
* if yes, then we _can't_ `$(PY_ENV)/requirements.sha256` : each invocation of `pyenv` will reinstall the virtualenv.

File: utils/makefile.python (hunk 4 above)
Comment (return42) [reply utils/makefile.python:119]: Sorry I am an idiot .. I totally forgot the main problem

File: utils/makefile.python (hunk 3 above)
Comment (return42) [reply utils/makefile.python:139]: These makefiles has been implemented for general use cases.  I use them in several projects as [boilerplate](https://return42.github.io/boilerplate/slides/index.html#/).  The motivation was to have  typical development workflows localy at hand, build on simple tools available nearly everywhere.

Back to your question _why there are two targets ?_ 

pyenv: builds the environment for development / typically `requirements.txt`
penvinstall: installs the python package in this environment (`pip install -e`) / typically `setup.py` 

In the searx project `requirements.txt` and `setup.py` include identical _requirements_ but this is [not typical for python projects](https://github.com/searx/searx/issues/492#issuecomment-768923721).  Beside this you can also think about projects containing more than one package.

> Is there a case where the pyenv target is invoked but not the pyenvinstall target ?

With the words above, my answer is: _it depends .._ But from `makefile.python`'s POV the answer is : yes

Your questions helped me to sort my thoughts, but for today I am to tired .. tomorrow I have more time, then I'll think about it again ..

- pyyaml in setup.py
- _each invocation of pyenv will reinstall the virtualenv_

File: utils/makefile.python (hunk 4 above)
Comment (return42) [reply utils/makefile.python:119]: Hm .. after thinking again .. pyyaml  should already be installed when the virtualenv is build

https://github.com/dalf/searx/blob/fe3085d68e965732ca7291a14225b34a7e046e92/utils/makefile.python#L122-L129

and this will be installed before pyenvinstall is called:

https://github.com/dalf/searx/blob/fe3085d68e965732ca7291a14225b34a7e046e92/utils/makefile.python#L193

In https://github.com/searx/searx/pull/2517#discussion_r567049019 I described it:

> pyenv: builds the environment for development / typically `requirements.txt`
> penvinstall: installs the python package in this environment (`pip install -e`) / typically `setup.py` 

The other point you asked:

> each invocation of pyenv will reinstall the virtualenv

Our mistake was to build `sha256sum requirements*.txt > $(PY_ENV)/requirements.sha256 ;\` in ` cmd_pyenvinstall` .. this optimizing should take place in `cmd_virtualenv`.

Or I am wrong? .. I will give it a try ...

Comment (dalf) [reply utils/makefile.python:119]: > Hm .. after thinking again .. pyyaml should already be installed when the virtualenv is build

cmd_virtualenv seems to fail sometimes on github action:
* in this case, the ./local contains an initialized virtualenv, but without pyyaml
* cmd_virtualenv won't detect that pyyaml is missing (or other dependencies): the virtualenv exists, `...python` is an executable.
* then cmd_pyenvinstall fails because pyyaml is missing.
* the cache is updated with a broken virtualenv.
* all new builds on github action fail.

An alternative implementation (without `...grep pyyaml ./requirements.txt)...` ) :
* after `$(PY_ENV_BIN)/python -m pip install $(PIP_VERBOSE) -r requirements.txt;`, a file is created (or just a copy of requirements.txt)
* `if [ -d "./$(PY_ENV)" -a -x "./$(PY_ENV_BIN)/python" ]; then` also check if this file exists.
* basically it duplicates `if ! cat $(PY_ENV)/requirements.sha256 2>/dev/null | sha256sum --check --status 2>/dev/null; then` in cmd_virtualenv.

Comment (return42) [reply utils/makefile.python:119]: I work on a solution .. give me some time this day and I will show you my solution:

> An alternative implementation 
> ...
> ..

All this goes in the right direction to hardening, but instead hardening the installation, we have to hardening the venv build .. I will consider it in my solution / thanks!

Comment (return42) [reply utils/makefile.python:119]: @dalf I implemented a alternative PR #2524 ... its all described there ..
